### PR TITLE
Transformer

### PR DESCRIPTION
**Pull request by zhangguanheng66**
Create a PR for comments. The model is still WIP but I want to have some feedbacks before moving too far. The transformer model depends on several modules, like MultiheadAttention (landed), PositionalEncoding, FeedForward... If they are available already, I will switch to the existing ones. Otherwise, those modules could be landed by separate PRs.

Transformer is implemented based on the paper (https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf). Users have the flexibility to build a transformer with self-defined and/or built-in components (i.e encoder, decoder, encoder_layer, decoder_layer). Calling buildTransformerModel() function will generate a generic transformer model. Then, Users could modify sub-layers as needed.

A test case was created to train a simple transformer model (see TestNN.test_transformer_number_match).

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
@cpuhrsch @soumith @myleott. Any feedbacks are appreciated.

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+# Temporarily leave Embeddings module here. Will be moved somewhere else.
+class Embeddings(nn.Module):
+	def __init__(self, d_model, vocab):
+		super(Embeddings, self).__init__()
+		self.lut = nn.Embedding(vocab, d_model)
+		self.d_model = d_model
+
+	def forward(self, x):
+		return self.lut(x) * math.sqrt(self.d_model)
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: I think it's fine if you move this explicitly into the [callsites](https://github.com/pytorch/pytorch/pull/19274/files#diff-c5db6ae353758a64f3e73dd758728cb7R15) instead of creating another layer, since self.d_model is only a scalar.

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+class FeedForward(nn.Module):
+	def __init__(self, d_model, d_ff=2048, dropout = 0.1):
+		super(FeedForward, self).__init__() 
+		# We set d_ff as a default to 2048
+		self.linear1 = nn.Linear(d_model, d_ff)
+		self.dropout = nn.Dropout(dropout)
+		self.linear2 = nn.Linear(d_ff, d_model)
+	
+	def forward(self, x):
+		return self.linear2(self.dropout(F.relu(self.linear1(x))))
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: This is a very common pattern in general. For now I think it's ok if you do this explicitly instead of using a separate Module within Encoder and DecoderLayer and comment that it could be fused.

**Comment by cpuhrsch** [comment]
Overall I think the temporary layers should be made private or merged into the respective models (and then marked as fuse-able). That'll minimize the number of potential additions to our nn library to only the Transformer components. After we  the tests pass and we have more docs we should pull in more people to see if we can get them to use it within their libraries. If this diff becomes rather contentious, we should consider merging it into [torchtext](https://github.com/pytorch/text) first.

**Comment by cpuhrsch** [comment]
cc @soumith

**Comment by cpuhrsch** [comment]
cc @gchanan

File: test/test_nn.py [hunk 3]
```diff
@@ -7699,6 +7701,39 @@ def test_adaptive_log_softmax(self):
...
+        V = 11
+        num_encoder_layers = 1
+        num_decoder_layers = 1
+        d_model = 16  # embedding
+        nhead = 2  # number of heads
+        dropout = 0.20
+        d_ff = 64
+        batch_size = 400
+        milliseconds = int(round(time.time() * 1000))
+        torch.manual_seed(milliseconds)
```
Comment (cpuhrsch) [inline test/test_nn.py]: Why is this necessary?

File: test/test_nn.py [hunk 4]
```diff
@@ -7699,6 +7701,39 @@ def test_adaptive_log_softmax(self):
...
+
+        model = Transformer(V, V, d_model=d_model, nhead=nhead, num_encoder_layers=num_encoder_layers, 
+                            num_decoder_layers=num_decoder_layers, d_ff=d_ff)
```
Comment (cpuhrsch) [inline test/test_nn.py]: What is "d_ff"? The name is a bit nondescript.

File: test/test_nn.py [hunk 5]
```diff
@@ -7699,6 +7701,39 @@ def test_adaptive_log_softmax(self):
...
+        for epoch in range(10):
+            model.train()
+            train_epoch(data_gen(V, batch_size, 21), model, criterion, model_opt) 
+            model.eval()
+            valid_epoch(data_gen(V, batch_size, 6), model, criterion) 
+
+        model.eval()
+
+        for _ in range(3):    
+            src = Variable(torch.randint(1, V, (10, 1)))
```
Comment (cpuhrsch) [inline test/test_nn.py]: I thought "Variable" was deprecated

File: test/test_nn.py [hunk 6]
```diff
@@ -7699,6 +7701,39 @@ def test_adaptive_log_softmax(self):
...
+            src[0][0] = 1
+            tgt = generate_test(model, src, max_len=10, start_symbol=1)
+            assert np.allclose(src, tgt, atol=1e-5)
```
Comment (cpuhrsch) [inline test/test_nn.py]: There's a torch "allclose" equivalent. You can't always assume np is available. If you have to rely on numpy, decorate the test to be skipped if numpy isn't available (see other tests for code examples).

File: test/test_nn.py [hunk 7]
```diff
@@ -8572,5 +8607,109 @@ def _buildEquivalentAffineTransforms3d(device, input_size, output_size, angle_ra
 # end TestNN.test_affine_* helpers
 
 
+# The following are some helpers for TestNN.test_transformer_number_match
+class DataBatch:
```
Comment (cpuhrsch) [inline test/test_nn.py]: If these are specific to your test, I'd add them to the local scope of that test, i.e. move them into the function itself. This way you avoid making these available to other tests and clearly specify the scope.

File: test/test_nn.py (hunk 3 above)
Comment (zhangguanheng66) [reply test/test_nn.py]: it is just a unit test for the transformer module.

File: test/test_nn.py [hunk 8]
```diff
@@ -8572,5 +8607,109 @@ def _buildEquivalentAffineTransforms3d(device, input_size, output_size, angle_ra
...
+    "Object for holding a batch of data with mask during training."
+    def __init__(self, src, trg=None, pad=0):
+        self.src = src.transpose(0, 1).contiguous()
+        self.src_mask = (src != pad).unsqueeze(-2)
+        if trg is not None:
+            self.trg = trg[:, :-1].transpose(0, 1).contiguous()
+            self.trg_y = trg[:, 1:].transpose(0, 1).contiguous()
+            self.ntokens = (self.trg_y != pad).data.sum().item()
+
+class NoamOpt:
```
Comment (cpuhrsch) [inline test/test_nn.py]: There should be an [optimizer](https://pytorch.org/docs/stable/optim.html) that already does this and if not, if this is common, we should add it explicitly. What features do our current optimizers lack, that this one provides?

File: test/test_nn.py [hunk 9]
```diff
@@ -8572,5 +8607,109 @@ def _buildEquivalentAffineTransforms3d(device, input_size, output_size, angle_ra
...
+        self._rate = rate
+        self.optimizer.step()
+
+    def rate(self, step=None):
+        if step is None:
+            step = self._step
+        return self.factor * (self.model_size ** (-0.5) * 
+                              min(step ** (-0.5), step * self.warmup ** (-1.5)))
+
+class LabelSmoothingLoss(nn.Module):
```
Comment (cpuhrsch) [inline test/test_nn.py]: If this is a common loss for Transformers, we should also add it to our losses.

File: test/test_nn.py (hunk 4 above)
Comment (zhangguanheng66) [reply test/test_nn.py]: Will make some changes in the next commit.

File: test/test_nn.py [hunk 10]
```diff
@@ -8572,5 +8607,109 @@ def _buildEquivalentAffineTransforms3d(device, input_size, output_size, angle_ra
...
+        true_dist.fill_(self.smoothing / (self.size - 2))
+        true_dist.scatter_(1, target.data.unsqueeze(1), self.confidence)
+        true_dist[:, self.padding_idx] = 0
+        mask = torch.nonzero(target.data == self.padding_idx)
+        if mask.dim() > 0:
+            true_dist.index_fill_(0, mask.squeeze(), 0.0)
+        self.true_dist = true_dist
+        return self.criterion(x, Variable(true_dist, requires_grad=False))
+
+def train_epoch(data_iter, model, criterion, opt):
```
Comment (cpuhrsch) [inline test/test_nn.py]: Since this is a common name, it's important to be local in scope to the test.

File: test/test_nn.py (hunk 5 above)
Comment (zhangguanheng66) [reply test/test_nn.py]: I don't see a warning for now.

File: test/test_nn.py [hunk 11]
```diff
@@ -7699,6 +7701,39 @@ def test_adaptive_log_softmax(self):
         out = asfm.predict(x)
         self.assertEqual(out, asfm.log_prob(x).argmax(dim=1))
 
+    def test_transformer_number_match(self):
```
Comment (cpuhrsch) [inline test/test_nn.py]: This types of integration tests are definitely useful. Are there smaller unit tests we could also look at to test for edge cases (empty input etc. etc.)?

File: test/test_nn.py (hunk 6 above)
Comment (zhangguanheng66) [reply test/test_nn.py]: I will search and make some changes.

File: test/test_nn.py [hunk 12]
```diff
@@ -7699,6 +7701,39 @@ def test_adaptive_log_softmax(self):
...
+        model = Transformer(V, V, d_model=d_model, nhead=nhead, num_encoder_layers=num_encoder_layers, 
+                            num_decoder_layers=num_decoder_layers, d_ff=d_ff)
+        model_opt = NoamOpt(model.encoder.src_embed.d_model, 1, 400,
+                            torch.optim.Adam(model.parameters(), lr=2.5, betas=(0.9, 0.98), eps=1e-9))
+        criterion = LabelSmoothingLoss(size=V, padding_idx=0, smoothing=0.0)
+
+        print("Training the model...")
+        for epoch in range(10):
+            model.train()
+            train_epoch(data_gen(V, batch_size, 21), model, criterion, model_opt)
```
Comment (cpuhrsch) [inline test/test_nn.py]: I'd move the Embedding outside of the Transformer. That is, the result of data_gen will be passed through an instance of nn.Embedding outside of this forward call. The reasoning for that is, that someone might want to use a different kind of Embedding and will then need to reimplement their own Transformer. We also don't do this for LSTM.

Comment (maxdel) [reply test/test_nn.py]: This is crucial for libs like allennlp or pytext for example.

File: torch/nn/modules/transformer.py [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
+
+    Examples::
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab)
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab, nhead=16, num_encoder_layers=12)
+    """
+
+    def __init__(self, src_vocab, tgt_vocab, d_model=512, nhead=8, num_encoder_layers=6,
+                 num_decoder_layers=6, d_ff=2048, dropout=0.1):
+        encoder_layer = TransformerEncoderLayer(d_model, nhead, d_ff, dropout)
+        src_embed = Embeddings(d_model, src_vocab)
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: I'd exclude this from the Transformer.

File: torch/nn/modules/transformer.py [hunk 14]
```diff
@@ -0,0 +1,519 @@
...
+    Examples::
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab)
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab, nhead=16, num_encoder_layers=12)
+    """
+
+    def __init__(self, src_vocab, tgt_vocab, d_model=512, nhead=8, num_encoder_layers=6,
+                 num_decoder_layers=6, d_ff=2048, dropout=0.1):
+        encoder_layer = TransformerEncoderLayer(d_model, nhead, d_ff, dropout)
+        src_embed = Embeddings(d_model, src_vocab) 
+        pos_encoder = PositionalEncoding(d_model, dropout)
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Same with the positional encoding.

File: torch/nn/modules/transformer.py [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+        tgt = tgt + self.dropout1(self.self_attn(tgt2, tgt2, tgt2, attn_mask=tgt_mask)[0])
+        tgt2 = self.norm2(tgt)
+        tgt = tgt + self.dropout2(self.multihead_attn(tgt2, memory, memory, attn_mask=memory_mask)[0])
+        tgt2 = self.norm3(tgt)
+        tgt = tgt + self.dropout3(self.ff(tgt2))        
+        return tgt
+
+
+# Temporarily leave LayerNorm module here. Will be moved somewhere else.
+class LayerNorm(Module):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: This should be replaceable by a [normalization layer](https://pytorch.org/docs/stable/nn.html#batchnorm1d).

File: torch/nn/modules/transformer.py [hunk 16]
```diff
@@ -0,0 +1,519 @@
...
+        dropout: the dropout value (default=0.1).
+
+    Examples::
+        >>> encoder_layer = nn.TransformerEncoderLayer(d_model, nhead)
+    """
+
+    def __init__(self, d_model, nhead, d_ff=2048, dropout=0.1):
+        super(TransformerEncoderLayer, self).__init__()
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.ff = FeedForward(d_model, d_ff=d_ff, dropout=dropout)
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: I suggest we inline the implementation of FeedForward until we're implementing a fused layer or such.

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: The reasoning is that FeedForward is quite small and we've historically decided to not introduce an abstraction for this type of model, but rather require the user to implement it herself.

File: torch/nn/modules/transformer.py [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+        Examples:
+            >>> output = layer_norm(x)
+        """
+        x_mean = x.mean(dim=-1, keepdim=True)
+        x_std = x.std(dim=-1, keepdim=True)
+        return self.alpha * (x - x_mean) / (x_std + self.eps) + self.bias
+
+
+# Temporarily leave FeedForward module here. Will be moved somewhere else.
+class FeedForward(Module):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: I suggest we remove this entirely.

File: torch/nn/modules/transformer.py [hunk 18]
```diff
@@ -0,0 +1,519 @@
...
+            >>> output = pos_encoder(x)
+        """
+
+        x = x + Variable(self.pe[:x.size(0), :], 
+                         requires_grad=False)
+        return self.dropout(x)
+
+
+# Temporarily leave Generator module here. Will be moved somewhere else.
+class Generator(Module):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: This also is very small and should be inlined.

File: torch/nn/modules/transformer.py [hunk 19]
```diff
@@ -0,0 +1,519 @@
...
+            output: [sequence length, batch size, embed dim]    
+
+        Examples:
+            >>> output = ff(x)
+        """
+        return self.linear2(self.dropout(F.relu(self.linear1(x))))
+
+
+# Temporarily leave Embeddings module here. Will be moved somewhere else.
+class Embeddings(Module):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: This should be left to the user.

File: torch/nn/modules/transformer.py [hunk 20]
```diff
@@ -0,0 +1,519 @@
...
+
+        Examples:
+            >>> output = embed_model(x)
+        """
+
+        return self.lut(x) * math.sqrt(self.d_model)
+
+
+# Temporarily leave PositionalEncoding module here. Will be moved somewhere else.
+class PositionalEncoding(Module):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: This should be left to the user. It's something you could define locally for your integration test.

File: torch/nn/modules/transformer.py [hunk 21]
```diff
@@ -0,0 +1,519 @@
...
+
+        memory = self.encode(src, src_mask)
+        output = self.decode(tgt, memory, tgt_mask, memory_mask)
+
+        if self.generator:
+            output = self.generator(output)
+
+        return output
+
+    def encode(self, src, src_mask=None):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Why is this method necessary?

File: torch/nn/modules/transformer.py [hunk 22]
```diff
@@ -0,0 +1,519 @@
...
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        output = src
+        if self.src_embed:
+            output = self.src_embed(output)
+
+        if self.pos_encoder:
+            output = self.pos_encoder(output)
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: You can assume that the user is passing the result of something akin to "self.src_embed(self.pos_encoder(src))" directly as the input (src). This is not something I'd add statically. We usually require the user to do the embedding herself and then pass in the result vector to our models.

File: torch/nn/modules/transformer.py [hunk 23]
```diff
@@ -0,0 +1,519 @@
...
+        self.decoder = decoder
+        self.generator = generator 
+
+
+class Transformer(TransformerBase):
+    r"""A transformer model applied for sequence-to-sequence transform. 
+        User is able to modified the attributes as needed.
+
+    Args:
+        src_vocab: the number of vocabularies in the source sequence (required).
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: If you remove the Embedding specialization you can call this input_size

File: test/test_nn.py (hunk 5 above)
Comment (gchanan) [reply test/test_nn.py]: it's deprecated -- we don't warn on some ubiquitous calls because basically all pytorch code uses them.

File: test/test_nn.py [hunk 24]
```diff
@@ -8572,5 +8607,109 @@ def _buildEquivalentAffineTransforms3d(device, input_size, output_size, angle_ra
...
+# The following are some helpers for TestNN.test_transformer_number_match
+class DataBatch:
+    "Object for holding a batch of data with mask during training."
+    def __init__(self, src, trg=None, pad=0):
+        self.src = src.transpose(0, 1).contiguous()
+        self.src_mask = (src != pad).unsqueeze(-2)
+        if trg is not None:
+            self.trg = trg[:, :-1].transpose(0, 1).contiguous()
+            self.trg_y = trg[:, 1:].transpose(0, 1).contiguous()
+            self.ntokens = (self.trg_y != pad).data.sum().item()
```
Comment (gchanan) [inline test/test_nn.py]: you probably don't need to call data anywhere here.

File: test/test_nn.py (hunk 3 above)
Comment (gchanan) [reply test/test_nn.py]: can you not use any of the existing test helpers in this file?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: torch/nn/modules/transformer.py (hunk 16 above)
Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: When you inline it, do mark it as such, so that we can possibly pull it back out later on

**Comment by sai-prasanna** [comment]
I remember fairseq's implementation has caching done for incremental decoding. Maybe something to look into ?

File: test/test_nn.py (hunk 5 above)
Comment (soumith) [reply test/test_nn.py]: yea, just `src = torch.randint(1, V, (10, 1))` is sufficient

Comment (soumith) [reply test/test_nn.py]: but again, see my comments above. a full convergence test is no place for our unit test suite, it does not catch implementation bugs.

File: test/test_nn.py (hunk 10 above)
Comment (soumith) [reply test/test_nn.py]: please remove before finalizing the PR (see my comments above)

File: test/test_nn.py (hunk 9 above)
Comment (soumith) [reply test/test_nn.py]: please remove before finalizing the PR (see my comments above)

File: torch/nn/modules/transformer.py (hunk 16 above)
Comment (soumith) [reply torch/nn/modules/transformer.py]: yea I agree.

File: torch/nn/modules/transformer.py (hunk 15 above)
Comment (soumith) [reply torch/nn/modules/transformer.py]: we have a native LayerNorm as well... https://pytorch.org/docs/stable/nn.html?highlight=layernorm#torch.nn.LayerNorm

File: torch/nn/modules/transformer.py (hunk 17 above)
Comment (soumith) [reply torch/nn/modules/transformer.py]: i concur

File: torch/nn/modules/transformer.py (hunk 19 above)
Comment (soumith) [reply torch/nn/modules/transformer.py]: agreed

**Review by soumith — CHANGES_REQUESTED** [review]

**Comment by zhangguanheng66** [comment]
I accidentally rebase and the commit log is too messy. Will create a new PR for merge request.

File: test/test_nn.py [hunk 25]
```diff
@@ -7806,6 +7807,58 @@ def test_adaptive_log_softmax(self):
         out = asfm.predict(x)
         self.assertEqual(out, asfm.log_prob(x).argmax(dim=1))
 
+    def test_transformer_number_match(self):
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_nn.py]: @soumith I agree that convergence tests might not be sufficient, but we have a `slowTest` decorator now; so if it only takes 20s to test for convergence, this seems like a good thing to put in as a slow test. What do you think?

[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_nn.py]: gradcheck here seems kind of pointless if we didn't actually implement any backwards explicitly (except, maybe, to check that the backwards is stable.)

[the remark this answers is not in the document]
Comment (soumith) [reply test/test_nn.py]: even as a slow-test, it's not useful. Even if it's buggy it will converge.
